File: examples/algorithms-go/graph.go
package algorithms

import (
	"container/heap"
	"math"
)

// BFS returns vertices in breadth-first search order starting from start.
// Time complexity: O(V + E). Space complexity: O(V).
func BFS(adj map[int][]int, start int) []int {
	if _, ok := adj[start]; !ok {
		return []int{}
	}
	visited := make(map[int]bool)
	queue := []int{start}
	visited[start] = true
	order := []int{}

	for len(queue) > 0 {
		v := queue[0]
		queue = queue[1:]
		order = append(order, v)
		for _, neighbor := range adj[v] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}
	return order
}

// BFSShortestPath returns the shortest path from start to goal using BFS.
// Returns nil if no path exists.
func BFSShortestPath(adj map[int][]int, start, goal int) []int {
	if start == goal {
		return []int{start}
	}
	if _, ok := adj[start]; !ok {
		return nil
	}
	visited := make(map[int]bool)
	visited[start] = true
	type node struct {
		vertex int
		path   []int
	}
	queue := []node{{start, []int{start}}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		for _, neighbor := range adj[curr.vertex] {
			if neighbor == goal {
				return append(append([]int{}, curr.path...), neighbor)
			}
			if !visited[neighbor] {
				visited[neighbor] = true
				newPath := append(append([]int{}, curr.path...), neighbor)
				queue = append(queue, node{neighbor, newPath})
			}
		}
	}
	return nil
}

// DFS returns vertices in depth-first search order starting from start.
// Time complexity: O(V + E). Space complexity: O(V).
func DFS(adj map[int][]int, start int) []int {
	if _, ok := adj[start]; !ok {
		return []int{}
	}
	visited := make(map[int]bool)
	stack := []int{start}
	visited[start] = true
	order := []int{}

	for len(stack) > 0 {
		v := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		order = append(order, v)
		// Push neighbors in reverse order for predictable order
		neighbors := adj[v]
		for i := len(neighbors) - 1; i >= 0; i-- {
			if !visited[neighbors[i]] {
				visited[neighbors[i]] = true
				stack = append(stack, neighbors[i])
			}
		}
	}
	return order
}

// HasPath checks if a path exists from start to goal.
func HasPath(adj map[int][]int, start, goal int) bool {
	if start == goal {
		return true
	}
	if _, ok := adj[start]; !ok {
		return false
	}
	visited := make(map[int]bool)
	stack := []int{start}
	visited[start] = true

	for len(stack) > 0 {
		v := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		for _, neighbor := range adj[v] {
			if neighbor == goal {
				return true
			}
			if !visited[neighbor] {
				visited[neighbor] = true
				stack = append(stack, neighbor)
			}
		}
	}
	return false
}

// item represents a node in the priority queue for Dijkstra.
type item struct {
	vertex   int
	distance int
	index    int
}

type priorityQueue []*item

func (pq priorityQueue) Len() int { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool {
	return pq[i].distance < pq[j].distance
}
func (pq priorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}
func (pq *priorityQueue) Push(x interface{}) {
	n := len(*pq)
	it := x.(*item)
	it.index = n
	*pq = append(*pq, it)
}
func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	it := old[n-1]
	it.index = -1
	*pq = old[:n-1]
	return it
}

// Dijkstra returns the shortest distances from start to all vertices.
// adj is a map from vertex to (neighbor, weight) pairs.
// Returns a map of vertex -> shortest distance.
func Dijkstra(adj map[int][][2]int, start int) map[int]int {
	dist := make(map[int]int)
	for v := range adj {
		dist[v] = math.MaxInt32
	}
	dist[start] = 0
	pq := make(priorityQueue, 0)
	heap.Init(&pq)
	heap.Push(&pq, &item{vertex: start, distance: 0})

	for pq.Len() > 0 {
		curr := heap.Pop(&pq).(*item)
		if curr.distance > dist[curr.vertex] {
			continue
		}
		for _, edge := range adj[curr.vertex] {
			neighbor, weight := edge[0], edge[1]
			newDist := curr.distance + weight
			if newDist < dist[neighbor] {
				dist[neighbor] = newDist
				heap.Push(&pq, &item{vertex: neighbor, distance: newDist})
			}
		}
	}
	return dist
}

File: examples/algorithms-js/graph.js
/**
 * graph.js - JavaScript图算法实现
 * 包含：BFS、DFS、最短路径判断
 */

class Graph {
    /**
     * BFS遍历
     * @param {Map<number, number[]>} adj - 邻接表
     * @param {number} start - 起始顶点
     * @returns {number[]} BFS顺序
     */
    static bfs(adj, start) {
        if (!adj.has(start)) return [];
        const visited = new Set();
        const queue = [start];
        const order = [];
        visited.add(start);
        while (queue.length > 0) {
            const v = queue.shift();
            order.push(v);
            for (const neighbor of (adj.get(v) || [])) {
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    queue.push(neighbor);
                }
            }
        }
        return order;
    }

    /**
     * DFS遍历
     */
    static dfs(adj, start) {
        if (!adj.has(start)) return [];
        const visited = new Set();
        const stack = [start];
        const order = [];
        visited.add(start);
        while (stack.length > 0) {
            const v = stack.pop();
            order.push(v);
            const neighbors = adj.get(v) || [];
            for (let i = neighbors.length - 1; i >= 0; i--) {
                const neighbor = neighbors[i];
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    stack.push(neighbor);
                }
            }
        }
        return order;
    }

    /**
     * 判断是否存在路径
     */
    static hasPath(adj, start, goal) {
        if (start === goal) return true;
        if (!adj.has(start)) return false;
        const visited = new Set();
        const stack = [start];
        visited.add(start);
        while (stack.length > 0) {
            const v = stack.pop();
            for (const neighbor of (adj.get(v) || [])) {
                if (neighbor === goal) return true;
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    stack.push(neighbor);
                }
            }
        }
        return false;
    }
}

// 简单测试
function assertEq(name, expected, actual) {
    const e = JSON.stringify(expected);
    const a = JSON.stringify(actual);
    if (e !== a) throw new Error(`FAIL ${name}: expected ${e}, got ${a}`);
}

function runTests() {
    const adj = new Map([
        [0, [1, 2]],
        [1, [2]],
        [2, [3]],
        [3, []]
    ]);
    assertEq("BFS", [0, 1, 2, 3], Graph.bfs(adj, 0));
    assertEq("DFS size", 4, Graph.dfs(adj, 0).length);
    assertEq("HasPath", true, Graph.hasPath(adj, 0, 3));
    assertEq("NoPath", false, Graph.hasPath(adj, 2, 0));
    console.log("All graph tests passed!");
}

runTests();

module.exports = { Graph };
